queue splitting guide (#337)
metalbear-co/mirrord.dev: content/docs/using-mirrord/queue-splitting.md

---
title: "Queue Splitting"
description: "Sharing queues by splitting messages between multiple clients and the cluster"
date: 2024-08-31T13:37:00+00:00
lastmod: 2024-08-31T13:37:00+00:00
draft: false
menu:
  docs:
    parent: "using-mirrord"
weight: 150
toc: true
tags: ["team", "enterprise"]
---

If your application consumes messages from a queue service, you might want the local application you're running with
mirrord to read messages from the queue. You probably don't want the local application to have to compete with the
remote one for messages, and you might also not want your local application to read all messages from the queue, as
that might disrupt the deployed application, or get in the way of teammates that are working on the same service.

> With _mirrord for Teams_, queue messages can be split between (possibly multiple) mirrord users, and the deployed

Reviewer: What does "possibly multiple" mean when qualifying a plural noun?

application.

> Users can set attribute filters in their mirrord configurations, and only get queue messages that have attributes

Reviewer: ```suggestion
Users can define a filter in their mirrord configurations, and their local app will only read queue messages that have match that filter
```

that match their filter.

> So far queue splitting is available for [Amazon SQS](https://aws.amazon.com/sqs/) queues. Pretty soon we'll support

Reviewer: Format this is a note, i.e.  `> **_NOTE:_**  The note content.`

Author: done in https://github.com/metalbear-co/mirrord.dev/pull/337/commits/63bf2867c4ebef86897a66e5e88324d745a22033

> So far queue splitting is available for [Amazon SQS](https://aws.amazon.com/sqs/) queues. Pretty soon we'll support

Reviewer: ```suggestion
So far queue splitting is available for [Amazon SQS](https://aws.amazon.com/sqs/). Pretty soon we'll support
```

> kafka queues as well.

Reviewer: ```suggestion
Kafka as well.
```

> 

Reviewer: How it works + Explanation + one or more diagrams

Author: Added in https://github.com/metalbear-co/mirrord.dev/pull/337/commits/c129fed5caf0bb1e9f6a851c13d31173c37411a0

> ## SQS-enabled mirrord operator installation

Reviewer: Enabling Queue Splitting in your Cluster

Author: https://github.com/metalbear-co/mirrord.dev/pull/337/commits/25c98fcb12bc0c84b89f14044d7a49d0e80397db

> In order to use the queue splitting feature, some extra values need be provided during the installation.

Reviewer: ```suggestion
In order to use the queue splitting feature, some extra values need be provided during the installation of the mirrord Operator.
```

First of all, the SQS splitting feature needs to be enabled.
When installing with the [mirrord-operator helm chart](https://github.com/metalbear-co/charts/tree/main/mirrord-operator)
this means setting the [`operator.sqsSplitting`](https://github.com/metalbear-co/charts/blob/61fec57ca913068a11f3dc8579bdaa377cb028a1/mirrord-operator/values.yaml#L22)
[value](https://helm.sh/docs/chart_template_guide/values_files/). When installing via the `mirrord operator setup`
command, the `--sqs-splitting` flag should be set.

> When sqs splitting is enabled during installation, some additional resources are created, and the SQS component of

Reviewer: ```suggestion
When SQS splitting is enabled during installation, some additional resources are created, and the SQS component of
```

> When sqs splitting is enabled during installation, some additional resources are created, and the SQS component of

Reviewer: Capitalize sqs throughout the post

> the mirrord operator is started.

Reviewer: Capitalize Operator throughout the post

Author: Also just "operator" when not occurring in "mirrord Operator" but by itself, like in "the Operator"? (Capitalizing in either of those cases seems weird to me)

Reviewer: In which cases, "operator" alone? Or also "mirrord Operator"?

Author: Both cases are a bit weird to me, but capitalizing "operator" when not preceded with mirrord I don't get, so I'm making sure you want those occurrences capitalized as well.

Reviewer: Don't worry about it

Additionally, the mirrord operator has to be able to create, read from, write to, and delete SQS queues.
If the queue messages are encrypted, the operator also needs the `kms:Encrypt`, `kms:Decrypt` and `kms:GenerateDataKey`
permissions.

For that, an IAM role with an appropriate policy has to be assigned to the operator's service account.
Please follow AWS's documentation on how to do that:
https://docs.aws.amazon.com/eks/latest/userguide/associate-service-account-role.html

> The ARN of the IAM role has to be passed when installing the operator. When installing with helm, the ARN is passed

Reviewer: Capitalize helm throughout the post

Author: done in https://github.com/metalbear-co/mirrord.dev/pull/337/commits/2c09a8c59b84c272c82c2809e9c5820b772f9f38

                    "author": "^me$",
                    "level": "^(beginner|intermediate)$"
                }
            }
        }
    }
}
```

* [`feature.split_queues`](/docs/reference/configuration/#feature-split_queues) is the configuration field you need
  to specify in order to filter queue messages. Directly under it, we have a mapping from a queue ID to a queue
  filter definition. This queue ID is the queue ID that was set in the
  [queue registry resource](#creating-a-queue-registry) of this target.
  * `message_filter` is a mapping from message attribute names to message attribute value regexes. Your local
    application will only see queue messages that have all the specified message attributes. In this case, the local
    application would only receive messages that have an attribute with the name "author" and the value "me", AND an
    attribute with the name "level" and one of the values "beginner" and "intermediate".

In the example above, a filter was only defined for one of the queues this target consumes. This is equivalent to
specifying a match-none filter for the second queue. Meaning our application will not see any messages in the
`ad-queue`.

> Once all users stop filtering a queue (stop their mirrord runs), the temporary SQS queues that mirrord created will be

Reviewer: ```suggestion
Once all users stop filtering a queue (i.e. end their mirrord sessions), the temporary SQS queues that mirrord created will be
```

---

> title: "Queue Splitting"

Reviewer: General note - there are a lot of long blocks of text here, when you look at the Cloudflare preview. Try breaking the text up a bit more with new paragraphs, bullets, etc. For example, if you're listing out options for Helm and `mirrord setup`, don't do it in the same paragraph.

Author: Better [now](https://github.com/metalbear-co/mirrord.dev/pull/337/commits/f59c7d204f455a79491c02553941f30e3b75fba4)?

description: "Sharing queues by splitting messages between multiple clients and the cluster"
date: 2024-09-10T13:37:00+00:00
lastmod: 2024-01-10T13:37:00+00:00
draft: false
images: []
linktitle: "Copy Target"
menu:
docs:
teams:
weight: 150
toc: true
tags: ["team", "enterprise"]
---

If your application consumes messages from a queue service, you might want the local application your running with

> mirrord to read messages from the queue. You probably don't want the local application to have to compete with the

Reviewer: I think this paragraph assumes some previous context. I would restructure it along the lines of:
"Two possible options are:
1. Simply running your application with mirrord without any special configuration. In this case, your local application will compete with the remote target for queue messages.
2. Running your application with copy-scaledown [link). This will result in..."

Another option is Queue Splitting...

Author: How is this: https://github.com/metalbear-co/mirrord.dev/pull/337/commits/5e318893484e4bddda52d4fafa42098ea26dd415

remote one for messages, and you might also not want your local application to read all messages from the queue, as
that might disrupt the deployed application, or get in the way of teammates that are working on the same service.

> With _mirrord for Teams_, queue messages can be split between (possibly multiple) mirrord users, and the deployed

Reviewer: No need to mention mfT - that's what the tags are for.
